### ISSUE TITLE
Feature Request: Undo Button

### ISSUE THREAD

**Issue by PinkWarrior** [mtgred/netrunner#1563]
People make the odd mistake all the time or miss click and although we have the commands to rectify these mistakes it can take sometime and newer players don't necessarly know the / commands. So an undo button for the last action you took I feel would be a welcome addition to jinteki.net

**Comment by JoelCFC25** [comment]
After some team discussion, it seems clear that supporting an automated undo feature would be a massive undertaking, making it very unlikely anytime in the foreseeable future. Manual adjustments, card dragging, and use of the console commands (easily accessible in the prominent Help tab at the top) can take care of everything.

**Comment by PinkWarrior** [comment]
Well that's certainly understandable. Thank you for the reply.

**Comment by zaroth** [comment]
in general, implementing the "action queue/history" (which is needed to implement Undo, see also https://en.wikipedia.org/wiki/Command_pattern )  could be really beneficial for debugging, keeping action structure organized, writing AI for the game, possibly some card implementations... but as mentioned, it needs a HUGE effort to refactor this. And huge refactors without similarly huge test coverage is NOT a recipe for success.

In short, for it to happen, we need:
- solid test coverage (current status for cards can be seen here https://docs.google.com/spreadsheets/d/1ICv19cNjSaW9C-DoEEGH3iFt09PBTob4CAutGex0gnE/pubhtml )
- a huge refactoring effort
- a brave soul willing to see it through.

(If you think you can fulfill the above criteria, apply to our team NOW! ;-) )

**Comment by erbridge** [comment]
There's an intermediary step here which is to have an stack for simple "reversible" actions like installing a card, clicking for a credit, or ending the turn. If a player takes one of those actions, we snapshot the game state before that action and offer an undo button to return the game to that snapshot. As soon as another player takes an action, or a non-reversible action is taken, the stack is cleared.
